### ISSUE TITLE
does not map to unicode

### ISSUE THREAD

**Issue by koala73** [vsespb/mt-aws-glacier#25]
Hello, 
it's working quite well, however, on certain directories, I get  a "does not map to unicode" error as below, any idea how to fix ?

Thanks

Call stack: 
App::MtAws::main(/root/mt-aws-glacier/./mtglacier:5)
App::MtAws::Journal::read_new_files(/root/mt-aws-glacier/lib/App/MtAws.pm:117)
App::MtAws::Journal::_read_files(/root/mt-aws-glacier/lib/App/MtAws/Journal.pm:203)
File::Find::find(/root/mt-aws-glacier/lib/App/MtAws/Journal.pm:244)
File::Find::_find_opt(/usr/share/perl/5.10/File/Find.pm:1297)
File::Find::_find_dir(/usr/share/perl/5.10/File/Find.pm:722)
App::MtAws::Journal::__ANON__(/usr/share/perl/5.10/File/Find.pm:902)
Encode::decode(/root/mt-aws-glacier/lib/App/MtAws/Journal.pm:243)
App::MtAws::**ANON**(/usr/lib/perl/5.10/Encode.pm:174)
Fatal Error: 0 utf8 "\xA0" does not map to Unicode at /usr/lib/perl/5.10/Encode.pm line 174.

**Comment by vsespb** [comment]
hi. what os, os version and filesystem ?
 23.02.2013 10:31 ÐÏÌØÚÏ×ÁÔÅÌØ "koala73" notifications@github.com ÎÁÐÉÓÁÌ:

**Comment by koala73** [comment]
Hello,

OS: Ubuntu 10.04.4 LTS
Filesystem : ext4

Let me know if I can help,

Thanks

**Comment by vsespb** [comment]
today's evening i will make a fix to produce more helpful error message, so
we will know what file cause this. but i can tell now that most likely this
file contain non utf8 character in filename which is kind of prohobited in
modern linuxes. this character looks like html's 'nbsp' in latin1 encoding.
 23.02.2013 13:12 ÐÏÌØÚÏ×ÁÔÅÌØ "koala73" notifications@github.com ÎÁÐÉÓÁÌ:

**Comment by koala73** [comment]
Possibly. The problem is without any kind of debug, it's hard to figure it out.
Appreciate your help !

**Comment by vsespb** [comment]
I released 0.87 beta. Now error message should look like this:

```
Fatal Error: 1 utf8 "\xA0" does not map to Unicode at /usr/local/lib/perl/5.10.1/Encode.pm line 174.
=== ===
ERROR: file/dir name [file name.dat] with invalid characters, inside directory [/dev/shm/mtaws/files/A01]
Dump of file name bytes:
SV = PV(0x22d2d08) at 0x2551190
  REFCNT = 1
  FLAGS = (POK,pPOK)
  PV = 0x25475e0 "file\240name.dat"\0
  CUR = 13
  LEN = 16
=== ===
```

note: `\240` is octal code of byte, same as `\xA0` hexadecimal

Small explanation about filename encodings:

On linux, filesystems/kernel treat filenames as bytes, not characters, so you can store the following as filenames:
1) valid UTF-8 characters
2) character in some other encodings (like Latin1 etc)
3) Not a characters, but just random bytes/garbage.

All userspace programs in Linux works fine only with (1).
If you set different (not default, not UTF8) locale, in theory they should work with (2), but I doubt.

Also some programs don't analyze filenames and don't really care about it's encoding, they will work with (2) and (3).

mtglacier currently works with (1) only. Soon should work with (2), but probably never with (3).

I cannot tell why you have '\xA0' in filename, looks like it's either (2) or (3). It depends on source of file.

But If you want to convert filenames from (2) to (1), there is a tool, called `convmv` (Ubuntu 10.04 package `convmv`) which can help converting filenames from one encoding to another or help find all filenames with broken encoding.

Let me know if this helps.

**Comment by koala73** [comment]
Thank you !

It works just fine

Regards

On Saturday, February 23, 2013, Victor Efimov wrote:

> I released 0.87 beta. Now error message should look like this:
> 
> Fatal Error: 1 utf8 "\xA0" does not map to Unicode at /usr/local/lib/perl/5.10.1/Encode.pm line 174.
> === ===
> ERROR: file/dir name [file name.dat] with invalid characters, inside directory [/dev/shm/mtaws/files/A01]
> Dump of file name bytes:
> SV = PV(0x22d2d08) at 0x2551190
>   REFCNT = 1
>   FLAGS = (POK,pPOK)
>   PV = 0x25475e0 "file\240name.dat"\0
>   CUR = 13
>   LEN = 16
> === ===
> 
> note: \240 is octal code of byte, same as \xA0 hexadecimal
> 
> Small explanation about filename encodings:
> 
> On linux, filesystems/kernel treat filenames as bytes, not characters, so
> you can store the following as filenames:
> 1) valid UTF-8 characters
> 2) character in some other encodings (like Latin1 etc)
> 3) Not a characters, but just random bytes/garbage.
> 
> All userspace programs in Linux works fine only with (1).
> If you set different (not default, not UTF8) locale, in theory they should
> work with (2), but I doubt.
> 
> Also some programs don't analyze filenames and don't really care about
> it's encoding, they will work with (2) and (3).
> 
> mtglacier currently works with (1) only. Soon should work with (2), but
> probably never with (3).
> 
> I cannot tell why you have '\xA0' in filename, looks like it's either (2)
> or (3). It depends on source of file.
> 
> But If you want to convert filenames from (2) to (1), there is a tool,
> called convmv (Ubuntu 10.04 package convmv) which can help converting
> filenames from one encoding to another or help found all filenames with
> broken encoding.
> 
> Let me know if this helps.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/vsespb/mt-aws-glacier/issues/25#issuecomment-13996028.

## 

_Elie Habib_
Jounieh, Lebanon
*M *+961 3810 690  *E *elie@anghami.com
